### PR TITLE
Fix issue with fragment field type conditions

### PR DESCRIPTION
**Pull request by sav007**
When fragment defined on the same type as parent field, don't generate fragment field type conditions otherwise it breaks parsing response for any new type.

See original issue for more details. 

Now for cases when fragment is defined on type of parent field we don't generate type conditions:
```
query TestQuery {
  hero { # hero here has type Character
    ...HeroDetails
  }  
}

fragment HeroDetails on Character {
   ...
}
```

https://github.com/apollographql/apollo-android/pull/2125/files#diff-988687da010965f3832fb8c9216ff862

Closes https://github.com/apollographql/apollo-android/issues/2097

### CODE REVIEW

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/response/RealResponseReader.java [hunk 1]
```diff
@@ -199,7 +199,7 @@ public RealResponseReader(Operation.Variables operationVariables, R recordSet,
       resolveDelegate.didResolveNull();
       result = null;
     } else {
-      CustomTypeAdapter<T> typeAdapter = scalarTypeAdapters.adapterFor(field.scalarType());
+      CustomTypeAdapter<T> typeAdapter = scalarTypeAdapters.adapterFor(field.getScalarType());
```
Comment (tasomaniac) [inline apollo-runtime/src/main/java/com/apollographql/apollo/internal/response/RealResponseReader.java]: It would be good to hold off for the depreciation changes in bug fixes. They are because of Kotlin changes and will make it harder to cherry pick into 1.4.x branch.

File: apollo-compiler/src/test/graphql/com/example/simple_fragment/TestQuery.java [hunk 2]
```diff
@@ -360,9 +360,7 @@ public int hashCode() {
 
       public static final class Mapper implements ResponseFieldMapper<Fragments> {
         static final ResponseField[] $responseFields = {
-          ResponseField.forFragment("__typename", "__typename", Arrays.<ResponseField.Condition>asList(
-            ResponseField.Condition.typeCondition(new String[] {"Human", "Droid"})
```
Comment (tasomaniac) [inline apollo-compiler/src/test/graphql/com/example/simple_fragment/TestQuery.java]: Just double checking. Is this intentional? I don't really understand what's going on here and Im curious how this works.

Comment (sav007) [reply apollo-compiler/src/test/graphql/com/example/simple_fragment/TestQuery.java:364]: Sure, I will try to explain.

Let's imagine that we have this query:
```
query TestQuery {
  hero { # hero here has type Character
    ...HeroDetails
  }  
}
fragment HeroDetails on Character {
   name
   ...
}
```

Here, `hero` field has interface type `Character` with `Human`, `Droid` implementation in schema v1. 
Now if fragment `HeroDetails` is defined on interface type `Character` (not on concrete type) we were generating meta field like:
```
ResponseField.forFragment("__typename", "__typename", Arrays.<ResponseField.Condition>asList(ResponseField.Condition.typeCondition(new String[] {"Human", "Droid"})
```
with `Human` and `Droid` type conditions, means that in order to read fragment from the response it MUST be on of these types.

Now the issue that with a new version of schema that introduces new implementation of interface type `Character` let's say `Mandalorian` we are going to fail to read fragment (even when it's defined on base interface type `Character`) because we locked ourselves and expect only `Human` or `Droid` types.

This PR fixes this issue by generating this meta field instead:
```
ResponseField.forFragment("__typename", "__typename", Collections.<ResponseField.Condition>emptyList())
```
basically we remove type conditions if fragment defined on interface or union type.

Comment (tasomaniac) [reply apollo-compiler/src/test/graphql/com/example/simple_fragment/TestQuery.java:364]: Thanks 👍
